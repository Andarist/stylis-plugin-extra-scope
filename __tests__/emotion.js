/**
 * @jest-environment jsdom
 */

import createCache from '@emotion/cache'
import { CacheProvider } from '@emotion/react'
import styled from '@emotion/styled'
import { render } from '@testing-library/react'
import React, { useState } from 'react'
import stylisExtraScope from '..'

const TestCacheProvider = ({ children }) => {
  const [emotionCache] = useState(() => {
    return createCache({
      key: 'extra-scope',
      stylisPlugins: [stylisExtraScope('#extraScopeContainer.extraScope')],
    })
  })

  return <CacheProvider value={emotionCache}>{children}</CacheProvider>
}

describe('StylisExtraScope with Emotion', () => {
  beforeEach(() => {
    document.querySelector('head').innerHTML = ''
    document.querySelector('body').innerHTML = ''
  })

  it('should handle simple input', () => {
    const Comp1 = styled.div({ fontSize: '10px' })

    render(
      <TestCacheProvider>
        <Comp1 />
      </TestCacheProvider>,
    )

    expect(document.documentElement).toMatchSnapshot()
  })

  it('should handle nested input', () => {
    const Comp1 = styled.div({ color: 'red' })
    const Comp2 = styled.div({ color: 'blue' })

    render(
      <TestCacheProvider>
        <Comp1>
          <Comp2 />
        </Comp1>
      </TestCacheProvider>,
    )

    expect(document.documentElement).toMatchSnapshot()
  })

  it('should handle @at-rules correctly', () => {
    const Comp1 = styled.div({
      color: 'blue',
      '@media': { fontSize: '1rem' },
      '@media (min-width: 768px)': { fontSize: '2rem' },
      '@keyframes': {
        '0%': {
          color: 'red',
        },
      },
    })

    render(
      <TestCacheProvider>
        <Comp1 />
      </TestCacheProvider>,
    )

    expect(document.documentElement).toMatchSnapshot()
  })

  it('should handle comma-separated selectors', () => {
    const Comp1 = styled.div({ color: 'blue', 'div, h1': { fontSize: '2rem' } })

    render(
      <TestCacheProvider>
        <Comp1 />
      </TestCacheProvider>,
    )

    expect(document.documentElement).toMatchSnapshot()
  })

  it('should handle parent selectors', () => {
    const Comp1 = styled.div({ color: 'red' })
    const Comp2 = styled.div({ '&&': { color: 'blue' } })

    render(
      <TestCacheProvider>
        <Comp1>
          <Comp2 />
        </Comp1>
      </TestCacheProvider>,
    )

    expect(document.documentElement).toMatchSnapshot()
  })

  it('should handle sibling selectors', () => {
    const Comp1 = styled.div({
      'div ~ div': { ':focus': { color: 'blue' } },
      '& ~ div': { fontSize: '1rem' },
      'div + div': { margin: '10px' },
      '& + div': { ':after': { color: 'red' } },
    })

    render(
      <TestCacheProvider>
        <Comp1 />
      </TestCacheProvider>,
    )

    expect(document.documentElement).toMatchSnapshot()
  })

  it('should handle child selectors', () => {
    const Comp1 = styled.div({
      '> .some-child-class': { ':focus': { color: 'blue' } },
      '& > .some-child-class': { fontSize: '1rem' },
      'div > .some-other-class': { margin: '10px' },
      '.some-parent-class > &': { ':after': { color: 'red' } },
    })

    render(
      <TestCacheProvider>
        <Comp1 />
      </TestCacheProvider>,
    )

    expect(document.documentElement).toMatchSnapshot()
  })

  it('should handle complicated child selectors', () => {
    const Comp1 = styled.div({
      '.class-one:focus ~ &, .class-two:focus > &': { outlineColor: 'red' },
    })

    render(
      <TestCacheProvider>
        <Comp1 />
      </TestCacheProvider>,
    )

    expect(document.documentElement).toMatchSnapshot()
  })

  it('should add single extra scope correctly for a rule declared after at rule', () => {
    const Comp1 = styled.div({
      '@media': { fontSize: '1rem' },
      color: 'blue',
    })

    render(
      <TestCacheProvider>
        <Comp1 />
      </TestCacheProvider>,
    )

    expect(document.documentElement).toMatchSnapshot()
  })
})
