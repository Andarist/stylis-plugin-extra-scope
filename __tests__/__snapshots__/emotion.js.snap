// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StylisExtraScope with Emotion should add single extra scope correctly for a rule declared after at rule 1`] = `
<html>
  <head>
    <style
      data-emotion="extra-scope"
      data-s=""
    >
      
      #extraScopeContainer.extraScope .extra-scope-t26pys{color:blue;}
    </style>
    <style
      data-emotion="extra-scope"
      data-s=""
    >
      
      @media{#extraScopeContainer.extraScope .extra-scope-t26pys{font-size:1rem;}}
    </style>
  </head>
  <body>
    <div>
      <div
        class="extra-scope-t26pys"
      />
    </div>
  </body>
</html>
`;

exports[`StylisExtraScope with Emotion should handle @at-rules correctly 1`] = `
<html>
  <head>
    <style
      data-emotion="extra-scope"
      data-s=""
    >
      
      #extraScopeContainer.extraScope .extra-scope-2zs2u8{color:blue;}
    </style>
    <style
      data-emotion="extra-scope"
      data-s=""
    >
      
      @media{#extraScopeContainer.extraScope .extra-scope-2zs2u8{font-size:1rem;}}
    </style>
    <style
      data-emotion="extra-scope"
      data-s=""
    >
      
      @media (min-width: 768px){#extraScopeContainer.extraScope .extra-scope-2zs2u8{font-size:2rem;}}
    </style>
    <style
      data-emotion="extra-scope"
      data-s=""
    >
      
      @keyframes{0%{color:red;}}
    </style>
  </head>
  <body>
    <div>
      <div
        class="extra-scope-2zs2u8"
      />
    </div>
  </body>
</html>
`;

exports[`StylisExtraScope with Emotion should handle child selectors 1`] = `
<html>
  <head>
    <style
      data-emotion="extra-scope"
      data-s=""
    >
      
      #extraScopeContainer.extraScope .extra-scope-1tlzw6p{}
    </style>
    <style
      data-emotion="extra-scope"
      data-s=""
    >
      
      #extraScopeContainer.extraScope .extra-scope-1tlzw6p &gt;.some-child-class{}
    </style>
    <style
      data-emotion="extra-scope"
      data-s=""
    >
      
      #extraScopeContainer.extraScope .extra-scope-1tlzw6p &gt;.some-child-class:focus{color:blue;}
    </style>
    <style
      data-emotion="extra-scope"
      data-s=""
    >
      
      #extraScopeContainer.extraScope .extra-scope-1tlzw6p&gt;.some-child-class{font-size:1rem;}
    </style>
    <style
      data-emotion="extra-scope"
      data-s=""
    >
      
      #extraScopeContainer.extraScope .extra-scope-1tlzw6p div&gt;.some-other-class{margin:10px;}
    </style>
    <style
      data-emotion="extra-scope"
      data-s=""
    >
      
      #extraScopeContainer.extraScope .some-parent-class&gt;.extra-scope-1tlzw6p{}
    </style>
    <style
      data-emotion="extra-scope"
      data-s=""
    >
      
      #extraScopeContainer.extraScope .some-parent-class&gt;.extra-scope-1tlzw6p:after{color:red;}
    </style>
  </head>
  <body>
    <div>
      <div
        class="extra-scope-1tlzw6p"
      />
    </div>
  </body>
</html>
`;

exports[`StylisExtraScope with Emotion should handle comma-separated selectors 1`] = `
<html>
  <head>
    <style
      data-emotion="extra-scope"
      data-s=""
    >
      
      #extraScopeContainer.extraScope .extra-scope-11ysuls{color:blue;}
    </style>
    <style
      data-emotion="extra-scope"
      data-s=""
    >
      
      #extraScopeContainer.extraScope .extra-scope-11ysuls div,#extraScopeContainer.extraScope .extra-scope-11ysuls h1{font-size:2rem;}
    </style>
  </head>
  <body>
    <div>
      <div
        class="extra-scope-11ysuls"
      />
    </div>
  </body>
</html>
`;

exports[`StylisExtraScope with Emotion should handle nested input 1`] = `
<html>
  <head>
    <style
      data-emotion="extra-scope"
      data-s=""
    >
      
      #extraScopeContainer.extraScope .extra-scope-tokvmb{color:red;}
    </style>
    <style
      data-emotion="extra-scope"
      data-s=""
    >
      
      #extraScopeContainer.extraScope .extra-scope-14ksm7b{color:blue;}
    </style>
  </head>
  <body>
    <div>
      <div
        class="extra-scope-tokvmb"
      >
        <div
          class="extra-scope-14ksm7b"
        />
      </div>
    </div>
  </body>
</html>
`;

exports[`StylisExtraScope with Emotion should handle parent selectors 1`] = `
<html>
  <head>
    <style
      data-emotion="extra-scope"
      data-s=""
    >
      
      #extraScopeContainer.extraScope .extra-scope-tokvmb{color:red;}
    </style>
    <style
      data-emotion="extra-scope"
      data-s=""
    >
      
      #extraScopeContainer.extraScope .extra-scope-rngag0{}
    </style>
    <style
      data-emotion="extra-scope"
      data-s=""
    >
      
      #extraScopeContainer.extraScope .extra-scope-rngag0.extra-scope-rngag0{color:blue;}
    </style>
  </head>
  <body>
    <div>
      <div
        class="extra-scope-tokvmb"
      >
        <div
          class="extra-scope-rngag0"
        />
      </div>
    </div>
  </body>
</html>
`;

exports[`StylisExtraScope with Emotion should handle sibling selectors 1`] = `
<html>
  <head>
    <style
      data-emotion="extra-scope"
      data-s=""
    >
      
      #extraScopeContainer.extraScope .extra-scope-1jdbxgg{}
    </style>
    <style
      data-emotion="extra-scope"
      data-s=""
    >
      
      #extraScopeContainer.extraScope .extra-scope-1jdbxgg div~div{}
    </style>
    <style
      data-emotion="extra-scope"
      data-s=""
    >
      
      #extraScopeContainer.extraScope .extra-scope-1jdbxgg div~div:focus{color:blue;}
    </style>
    <style
      data-emotion="extra-scope"
      data-s=""
    >
      
      #extraScopeContainer.extraScope .extra-scope-1jdbxgg~div{font-size:1rem;}
    </style>
    <style
      data-emotion="extra-scope"
      data-s=""
    >
      
      #extraScopeContainer.extraScope .extra-scope-1jdbxgg div+div{margin:10px;}
    </style>
    <style
      data-emotion="extra-scope"
      data-s=""
    >
      
      #extraScopeContainer.extraScope .extra-scope-1jdbxgg+div{}
    </style>
    <style
      data-emotion="extra-scope"
      data-s=""
    >
      
      #extraScopeContainer.extraScope .extra-scope-1jdbxgg+div:after{color:red;}
    </style>
  </head>
  <body>
    <div>
      <div
        class="extra-scope-1jdbxgg"
      />
    </div>
  </body>
</html>
`;

exports[`StylisExtraScope with Emotion should handle simple input 1`] = `
<html>
  <head>
    <style
      data-emotion="extra-scope"
      data-s=""
    >
      
      #extraScopeContainer.extraScope .extra-scope-9c7r58{font-size:10px;}
    </style>
  </head>
  <body>
    <div>
      <div
        class="extra-scope-9c7r58"
      />
    </div>
  </body>
</html>
`;
